Pubsub Subscription Support in Prow

sebastienvas opened the pull request:
Design Doc: https://goo.gl/gRsXkD

BenTheElder commented:
/cc @krzyzacy

krzyzacy commented:
/assign
/woof
will look shortly

stevekuznetsov commented:
Can you group the code to be reviewed in commits and the generated stuff separately? As it is it's a little hard to piece together the different relevant commits you have (squash "simplify code" commit for example)

sebastienvas commented:
@stevekuznetsov please take a look

sebastienvas commented on prow/pubsub/subscriber/server.go:
```diff
@@ -0,0 +1,230 @@
+				logrus.Warn("New config found, reloading pull Server")
+				// Making sure the current thread finishes before starting a new one.
+				errGroup.Wait()
+				// Starting a new thread with new config
```
What are the worker in your idea ? handlePulls ? By the way there is only one worker.
We still need to check if the config changed, and it is just easier to do it in the parent, such that we can cancel the child if a new config exist. Also I prefer passing the config down because from the parent since the parent is already doing a lookup. If the child is doing another lookup the parent and the child might not have the same config, and therefore the parent might issue an extra cancel to the child.

I considered keeping track of all the sub.Receive threads, and removing / adding them when the config is changed, but that s much more complex and prone to more bugs. I could not see the benefits of doing this over just restarting all the threads when a new config is here since pubsub will resend anyway if a message is not acked.

stevekuznetsov replied:
Hmmm. True, there is only one `handlePulls`. In one of my WIP pulls I have a config change subscription -- let me post that and we could use one impl.

stevekuznetsov replied:
I was thinking something like this https://github.com/kubernetes/test-infra/pull/10247

sebastienvas commented on prow/pubsub/subscriber/server.go:
```diff
@@ -0,0 +1,230 @@
+					msg.Ack()
+				})
+				if err != nil {
+					return err
```
done

sebastienvas commented on prow/config/config.go:
```diff
@@ -286,6 +289,9 @@ type Branding struct {
 	HeaderColor string `json:"header_color,omitempty"`
 }
 
+// PubsubSubscriptions holds Pubsub subscriptions we want to listen to.
```
done

sebastienvas commented on prow/cmd/sub/main.go:
```diff
@@ -0,0 +1,190 @@
+	"k8s.io/test-infra/prow/kube"
+
+	"golang.org/x/sync/errgroup"
+	"k8s.io/test-infra/prow/pubsub/subscriber"
```
done

sebastienvas commented on prow/cmd/sub/main.go:
```diff
@@ -0,0 +1,190 @@
+	flagOptions = &options{}
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	fs.StringVar(&flagOptions.masterURL, "masterurl", "", "URL to k8s master")
+	fs.StringVar(&flagOptions.kubeConfig, "kubeconfig", "", "Cluster config for the cluster you want to connect to")
```
yes

sebastienvas commented on prow/cmd/sub/main.go:
```diff
@@ -0,0 +1,190 @@
+	fs.BoolVar(&flagOptions.push, "push", false, "Using Push Server.")
+	fs.StringVar(&flagOptions.pushSecretFile, "push-secret-file", "", "Path to Pub/Sub Push secret file.")
+
+	fs.StringVar(&flagOptions.configPath, "config-path", "../../config.yaml", "Path to config.yaml.")
```
yes

krzyzacy commented on prow/cmd/crier/main.go:
```diff
@@ -132,61 +125,17 @@ func parseOptions() options {
 	return o
 }
 
-// TODO(krzyzacy): copy & paste, refactor this
```
thanks for doing this! :100:

krzyzacy commented on prow/pubsub/subscriber/metrics.go:
```diff
@@ -0,0 +1,87 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
babynit: 2018, and other places

sebastienvas replied:
Done

krzyzacy commented on prow/pubsub/subscriber/subscriper.go:
```diff
@@ -0,0 +1,177 @@
+/*
```
(filename: s/subscriper.go/subscriber.go)

sebastienvas replied:
done

krzyzacy commented on prow/pubsub/subscriber/subscriper.go:
```diff
@@ -0,0 +1,177 @@
+import (
+	"fmt"
+
+	"k8s.io/test-infra/prow/client/clientset/versioned"
```
(move this import down with other prow packages)

sebastienvas replied:
done

krzyzacy commented on prow/pubsub/subscriber/subscriper.go:
```diff
@@ -0,0 +1,177 @@
+	var prowJob kube.ProwJob
+	if r, err := extractPubSubMessage(msg.GetAttributes()); err != nil {
+		l.Warning("no pubsub information found to publish to")
+		prowJob = pjutil.NewProwJob(prowJobSpec, nil)
```
include all the labels from `periodicJob`, also below

sebastienvas replied:
done

krzyzacy commented on prow/pubsub/subscriber/subscriper.go:
```diff
@@ -0,0 +1,177 @@
+		}
+	}
+	if periodicJob == nil {
+		return fmt.Errorf("failed to find associated periodic job")
```
(I would also include the name in the log)

sebastienvas replied:
done

sebastienvas commented:
@krzyzacy please take a look. I fixed your comments, and updated to use the payload to pass in information. I also added an options to pass environment variable to a jobs which will be passed down to all containers (for the release use case since we might point tests to a new artifacts as an example)

krzyzacy commented on prow/kube/cluster.go:
```diff
@@ -0,0 +1,88 @@
+limitations under the License.
+*/
+
+package kube
```
so there's already https://github.com/kubernetes/test-infra/blob/master/prow/flagutil/kubernetes.go, maybe we should merge some logic, or it's time to unify all the different kubeclients we have? @stevekuznetsov

sebastienvas replied:
that's the horrible client that is not compatible with kubeconfig. I think we should kill this.

krzyzacy replied:
yeah, I mean, let everything use this :-)

krzyzacy commented:
/hold
for @sebastienvas to add some docs

sebastienvas commented:
@krzyzacy documentation added.

sebastienvas commented:
@krzyzacy please take a look.

krzyzacy commented:
rebase? :-)

krzyzacy commented:
/lgtm

sebastienvas commented:
/hold cancel

sebastienvas commented:
/approve

krzyzacy commented:
/approve
